chainimport/headers_import.go

```diff
@@ -0,0 +1,1816 @@
+		// They serve as sanity checks to catch unexpected
+		// inconsistencies.
+		if len(blockHeaders) != len(filterHeaders) {
+			return fmt.Errorf("mismatch between block headers "+
+				"(%d) and filter headers (%d) for batch %d-%d",
+				len(blockHeaders), len(filterHeaders),
+				batchStart, batchEnd)
+		}
+		if len(blockHeaders) == 0 {
+			return fmt.Errorf("no headers read for "+
+				"batch %d-%d", batchStart, batchEnd)
+		}
+
+		setLastFilterHeaderHash(filterHeaders, blockHeaders)
+
+		err := s.writeHeadersToTargetStores(
+			blockHeaders, filterHeaders, batchStart, batchEnd,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to write headers to "+
+				"target stores: %v", err)
+		}
+
+		log.Debugf("Wrote headers batch from height "+
+			"%d to %d (%d headers)", batchStart, batchEnd,
+			batchEnd-batchStart+1)
+
+		// Move to next batch.
+		batchStart = batchEnd + 1
+	}
+
+	log.Infof("Successfully added %d new headers from heights %d to %d",
+		totalHeaders, startHeight, endHeight)
+
+	return nil
+}
+
+// Write block and filter headers to the target stores in a specific order to
+// ensure proper rollback operations if needed. It also rollbacks any headers
+// written if any to target stores incase of any failures.
+func (s *HeadersImport) writeHeadersToTargetStores(
+	blockHeaders []headerfs.BlockHeader,
+	filterHeaders []headerfs.FilterHeader,
+	batchStart, batchEnd uint32) error {
+
+	// Write block headers batch to target store.
+	if err := s.options.TargetBlockHeaderStore.WriteHeaders(
+		blockHeaders...,
+	); err != nil {
+		return fmt.Errorf("failed to write block headers "+
+			"batch %d-%d: %w", batchStart, batchEnd, err)
+	}
+
+	// Write filter headers batch to target store.
+	if err := s.options.TargetFilterHeaderStore.WriteHeaders(
+		filterHeaders...,
+	); err != nil {
+		// If we've reached here, the headers failed to be written to
+		// target filter store because of I/O errors regarding binary
+		// file or filter db store, and it is automatically rolled back
+		// upstream.
+		//
+		// The whole import operation needs to satisfy the conjunction
+		// property for both block and filter header stores - it's all
+		// or nothing, so they need to be at the same length. We must
+		// rollback the block headers to maintain this consistency.
+		blkStore := s.options.TargetBlockHeaderStore
+		blockHeadersToTruncate := uint32(len(blockHeaders))
+		_, rollbackErr := blkStore.RollbackBlockHeaders(
+			blockHeadersToTruncate,
+		)
+		if rollbackErr != nil {
+			return fmt.Errorf("failed to rollback %d headers from "+
+				"target block header store after filter "+
+				"headers write failure. Block error: %w, "+
+				"filter error: %v", blockHeadersToTruncate,
+				rollbackErr, err)
+		}
+
+		return fmt.Errorf("failed to write filter headers "+
+			"batch %d-%d: %w", batchStart, batchEnd, err)
+	}
+
+	return nil
+}
+
+// ImportOptions defines parameters for the import process.
+type ImportOptions struct {
```

Comment:
Hmm, I'd expect that we pass in these import option to the importer itself, which then uses them to import. Right now we have that the import options is what we execute against directly to import.